rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,678 @@
+      - (-) Requires a Nix language version bump to implement this RFC
+
+   </details>
+
+1. Each time the language specification (currently as embodied by the Nix language evaluator) is changed, the language version must be incremented.
+
+   <details><summary>Arguments</summary>
+
+     - (+) The principled solution: guarantees reproducibility given a fixed language version
+     - (-) Implies additional overhead in development effort:
+       - Either for Nix maintainers to accommodate that practice in the release lifecycle
+         - For example, one would have to batch language changes for a version bump to limit the number of increments
+           - (+) This would be beneficial for alternative implementations in terms of churn and effort to keep up
+       - Or implementors of alternative evaluators catching up with changes
+         - (+) Specifying the language precisely via the version actually offers alternative implementations an alternative to catching up: only support a given language version
+     - (+) This essentially nudges one to organise Nix language (specification or evaluator implementation ) development to be more independent of the rest of Nix
+       This is good, since it in turn forces stronger separation of concerns and more architectural clarity
+     - (-) Prohibits best-effort attempts at evaluating expressions with possibly incompatible evaluators
+       - (+) With the proposed level of strictness, one doesn't have to rely on best effort but can instead be explicit
+     - (-) Fixing evaluator bugs (i.e., clearly unintended behavior) after releases would technically require a version bump and therefore (theoretically) cooperation by expression authors
+       - This could be communicated with an increment in the Nix patch-level version, as is already practice
+
+   </details>
+
+   <details><summary>Alternatives</summary>
+
+   - Bump version only when evaluation result on prior version evaluators would be *substantially* different
+     - (+) Leaves room for judgement by developers
+       - (+) Allows controlling progression of versions to some degree
+     - (-) Conversely, leaves room for sneaking in breaking changes unannounced
+       - (-) This loses compatibility guarantees we'd get from a stricter paradigm
+       - (-) Deprives expression authors of ability to be selective with evaluator versions
+     - (-) Due to hashing this is often not much different from taking *any* change into account
+   - Bump version when prior evaluators would fail
+     - (-) Derivation hashes may change between evaluator versions if different evaluation results were allowed within one version
+       - This amounts to giving up on specifying the language exhaustively using a version label
+     - (+) Fewer version increments or precautions required given current development practice
+
+   </details>
+
+1. Whenever Nix drops support for evaluating prior language versions, a major version bump is required.
```

Comment:
I think this RFC should decide whether it wants to treat the Nix language and `nix` as coupled, or to be just about the Nix language independent of implementation. In the latter case, implementation specific things should be moved to "examples and interactions".